File: dbscan/fuzzy_dbscan.py
import numpy as np
import pandas as pd

# This function executes a fuzzy dbscan algorithm.
# Parameters are:
# eps:            epsilon distance
# minPtsMin:      minimum amount of points to be in the neighborhood of a
#                 data point p for p to be recognized as a core point.
# mintPtsMax:     maximum amount a points in the neighborhood of a data point
#                 which leads to maximum membership degree of 1 for points with
#                 at least minPtsMax neighbors. This parameter helps to recognize
#                 more degrees of density. Thats's why it is recommended to use
#                 big values.
# data:           A numpy.ndarray with data points as rows and columns as attributes
#
# Returns a vector of cluster ids, one cluster id for each data point.

def FuzzyDBSCAN(data, eps, minPtsMin, minPtsMax):
    distances = computeDistance(data)
    numPoints = data.shape[0]
    currentCluster = -1
    memberships = [[-1] for i in range(numPoints)]
    visited = [False] * numPoints
    for i in range(numPoints):
        if(visited[i]):
            continue

        neighbors = computeNeighbors(distances, i, eps)

        # if data point is core point
        if(len(neighbors) > minPtsMin):
            visited[i] = True
            currentCluster = currentCluster + 1
            if currentCluster > 0:
                for row in memberships:
                    row.append(-1)
            
            expandFuzzyCluster(i, neighbors, eps, minPtsMin, minPtsMax, visited, memberships, distances, currentCluster)
    
    clustering = []
    for i in range(numPoints):
        cluster = -1
        maxMembership = -1 
        for j in range(currentCluster+1):
            currentMembership = memberships[i][j]
            if(currentMembership > maxMembership):
                cluster = j
                maxMembership = currentMembership
        clustering.append(cluster)
    return clustering 
            
def expandFuzzyCluster(point, neighbors, eps, minPtsMin, minPtsMax, visited, memberships, distances, currentCluster):
    borderPoints = set()
    corePoints = set()
    memberships[point][currentCluster] = computeMembershipDegree(len(neighbors), minPtsMin, minPtsMax)
    corePoints.add(point)
    while neighbors:
        i = neighbors.pop()
        if not visited[i] and not(i in borderPoints):
            neighbors2 = computeNeighbors(distances, i, eps)
            visited[i] = True
            corePoints.add(i)
            neighbors = neighbors.union(neighbors2)
            memberships[i][currentCluster] = computeMembershipDegree(len(neighbors2), minPtsMin, minPtsMax)
        else:
            borderPoints.add(i)
    while borderPoints:
        i = borderPoints.pop()
        neighbors2 = computeNeighbors(distances, i, eps)
        coreNeighbors = neighbors2.intersection(corePoints)
        biggestMembership = -1
        while coreNeighbors:
            j = coreNeighbors.pop()
            currentMembership = memberships[j][currentCluster]
            biggestMembership = max(biggestMembership, currentMembership)
        
        memberships[i][currentCluster] = biggestMembership
    return

# Function to compute the eps-neighborhood of a data point as a set of indizes.
#
# Parameters are:
# distances:    numpy.ndarray that is an upper triangular matrix with diagonal 0-entries.
# point:        Index in distance matrix of data point to compute the neighborhood for.
# eps:          Value to define distance of epsilon neighborhood epsilon.
#
# Returns set of neighbor points as indizes into distance matrix.
def computeNeighbors(distances, point, eps):
    neighbors = set()

    for i in range(point):
        if distances[i][point] <= eps:
            neighbors.add(i)
    
    numPoints = distances.shape[1]
    for i in range(point, numPoints):
        if distances[point][i] <= eps:
            neighbors.add(i)

    return neighbors

# Function to compute the membership degree of a data point to a cluster.
def computeMembershipDegree(numNeighbors, minPtsMin, minPtsMax):
    if(numNeighbors >= minPtsMax):
        return 1
    if(numNeighbors <= minPtsMin):
        return 0
    return float(numNeighbors - minPtsMin) / float(minPtsMax - minPtsMin)

# Function to compute the Euclidean distance matrix
def computeDistance(data):
    lenArrayOfPoints = len(data)
    dimension = len(data[0])

    distanceMatrix = []
    for i in range(lenArrayOfPoints):
        distanceCollector = [[]]
        for j in range(lenArrayOfPoints):
            if i >= j: 
                distanceCollector[0].extend([0])
            else:
                euclideanDistanceAddition = 0
                for k in range(dimension):
                    euclideanDistanceAddition = euclideanDistanceAddition + (data[i][k] - data[j][k])**2
                euclideanDistanceAddition = euclideanDistanceAddition**.5
                distanceCollector[0].extend([euclideanDistanceAddition])
        distanceMatrix.extend(distanceCollector)
    distanceMatrix = np.array(distanceMatrix)
    return distanceMatrix

data = pd.read_csv('/home/ankhanh/learning/data_mining/data/sample_dataset.csv')
new_data = np.array(data.iloc[:, 2:4])
distanceMatrix = computeDistance(new_data)
# print(repr(distanceMatrix[1]))
clustering = FuzzyDBSCAN(new_data, 10, 2, 5)
for clusterid in clustering:
        print(clusterid, end = ' ')

File: dbscan_cpp/fuzzy_dbscan.cpp
#include <iostream>
#include <vector>
#include <cmath>
#include <iostream>
#include <set>
#include <fstream>
#include <string>
#include <sstream>
//#include <bits/stdc++.h>
using namespace std;

// Function to compute the Euclidean distance matrix
vector<vector<double>> computeDistance(vector<vector<double>> data) {
    int lenArrayOfPoints = data.size();
    int dimension = data[0].size();

    vector<vector<double>> distanceMatrix(lenArrayOfPoints, vector<double>(lenArrayOfPoints, 0));
    for (int i = 0; i < lenArrayOfPoints; i++) {
        for (int j = i + 1; j < lenArrayOfPoints; j++) {
            double euclideanDistanceAddition = 0;
            for (int k = 0; k < dimension; k++) {
                euclideanDistanceAddition += pow(data[i][k] - data[j][k], 2);
            }
            euclideanDistanceAddition = sqrt(euclideanDistanceAddition);
            distanceMatrix[i][j] = euclideanDistanceAddition;
            distanceMatrix[j][i] = euclideanDistanceAddition;
        }
    }
    return distanceMatrix;
}
double distance(vector<double> a, vector<double> b) {
    double euclideanDistanceAddition = 0;
    for (int k = 0; k < a.size(); k++) {
        euclideanDistanceAddition += pow(a[k] - b[k], 2);
    }
    euclideanDistanceAddition = sqrt(euclideanDistanceAddition);
    return euclideanDistanceAddition;
}

// Function to compute the membership degree of a data point to a cluster.
double computeMembershipDegree(int numNeighbors, int minPtsMin, int minPtsMax) {
    if (numNeighbors >= minPtsMax) {
        return 1;
    }
    if (numNeighbors <= minPtsMin) {
        return 0;
    }
    return double(numNeighbors - minPtsMin) / double(minPtsMax - minPtsMin);
}

// Function to compute the eps-neighborhood of a data point as a set of indizes.
//
// Parameters are:
// distances:    vector<vector<double>> that is an upper triangular matrix with diagonal 0-entries.
// point:        Index in distance matrix of data point to compute the neighborhood for.
// eps:          Value to define distance of epsilon neighborhood epsilon.
//
// Returns set of neighbor points as indizes into distance matrix.
vector<int> computeNeighbors(vector<vector<double>> distances, int point, double eps) {
    vector<int> neighbors;

    for (int i = 0; i < point; i++) {
        if (distances[i][point] <= eps) {
            neighbors.push_back(i);
        }
    }

    int numPoints = distances.size();
    for (int i = point; i < numPoints; i++) {
        if (distances[point][i] <= eps) {
            neighbors.push_back(i);
        }
    }

    return neighbors;
}

void expandFuzzyCluster(int point, vector<int> neighbors, double eps, int minPtsMin, int minPtsMax, vector<bool>& visited, vector<vector<double>>& memberships, vector<vector<double>> distances, int& currentCluster) {
    set<int> borderPoints;
    set<int> corePoints;
    memberships[point][currentCluster] = computeMembershipDegree(neighbors.size(), minPtsMin, minPtsMax);
    corePoints.insert(point);
    while (!neighbors.empty()) {
        int i = neighbors.back();
        neighbors.pop_back();
        if (!visited[i] && !borderPoints.count(i)) {
            vector<int> neighbors2 = computeNeighbors(distances, i, eps);
            visited[i] = true;
            corePoints.insert(i);
            neighbors.insert(neighbors.end(), neighbors2.begin(), neighbors2.end());
            memberships[i][currentCluster] = computeMembershipDegree(neighbors2.size(), minPtsMin, minPtsMax);
        }
        else {
            borderPoints.insert(i);
        }
    }
    while (!borderPoints.empty()) {
        int i = *borderPoints.begin();
        borderPoints.erase(borderPoints.begin());
        vector<int> neighbors2 = computeNeighbors(distances, i, eps);
        set<int> coreNeighbors;
        for (int j : neighbors2) {
            if (corePoints.count(j)) {
                coreNeighbors.insert(j);
            }
        }
        double biggestMembership = -1;
        for (int j : coreNeighbors) {
            double currentMembership = memberships[j][currentCluster];
            biggestMembership = max(biggestMembership, currentMembership);
        }
        memberships[i][currentCluster] = biggestMembership;
    }
}

vector<int> FuzzyDBSCAN(vector<vector<double>> data, double eps, int minPtsMin, int minPtsMax) {
    vector<vector<double>> distances = computeDistance(data);
    int numPoints = data.size();
    int currentCluster = -1;
    vector<vector<double>> memberships(numPoints, vector<double>(1, -1));
    vector<bool> visited(numPoints, false);
    for (int i = 0; i < numPoints; i++) {
        if (visited[i]) {
            continue;
        }

        vector<int> neighbors = computeNeighbors(distances, i, eps);

        // if data point is core point
        if (neighbors.size() > minPtsMin) {
            visited[i] = true;
            currentCluster++;
            if (currentCluster > 0) {
                for (auto& row : memberships) {
                    row.push_back(-1);
                }
            }

            expandFuzzyCluster(i, neighbors, eps, minPtsMin, minPtsMax, visited, memberships, distances, currentCluster);
        }
    }

    vector<int> clustering(numPoints, -1);
    for (int i = 0; i < numPoints; i++) {
        int cluster = -1;
        double maxMembership = -1;
        for (int j = 0; j <= currentCluster; j++) {
            double currentMembership = memberships[i][j];
            if (currentMembership > maxMembership) {
                cluster = j;
                maxMembership = currentMembership;
            }
        }
        clustering[i] = cluster;
    }
    return clustering;
}

int main() {
    // Example usage
    // vector<vector<double>> data = {{1, 2, 3}, {4, 5, 6}, {7, 8, 9}};
    ifstream file("C:\\Users\Admin\source\repos\akBear23\K-DBSCAN\data\sample_dataset.csv");
    if (!file.is_open()) {
        cout << "Error opening file" << endl;
        return 1;
    }
    vector<vector<double>> data;
    string line = "";
    getline(file, line);
    line = "";
    while (getline(file, line)) {
        vector<double> row;
        string feature_type;
        int id; 
        double loc_x, loc_y;
        double value;
        string tmp_str="";
        stringstream ss(line);
        
        getline(ss, feature_type, ',');
        getline(ss, tmp_str, ',');
        id = stoi(tmp_str);
        tmp_str="";
        getline(ss, tmp_str, ',');
        loc_x = stod(tmp_str);
        tmp_str="";
        getline(ss, tmp_str, ',');
        loc_y = stod(tmp_str);
        row.push_back(loc_x);
        row.push_back(loc_y);
        // printf("%f %f\n", loc_x, loc_y);
        data.push_back(row);
    }

    // Close the file
    file.close();

    double eps = 10;
    int minPtsMin = 2;
    int minPtsMax = 5;

    // cout << data[0].size() << endl;
    vector<vector<double>> distaceMatrix = computeDistance(data);
    // for (int i = 0; i < distaceMatrix.size(); i++) {
    //     for (int j = 0; j < distaceMatrix[i].size(); j++) {
    //         cout << distaceMatrix[i][j] << " ";
    //     }
    //     cout << endl;
    // }
    vector<int> clustering = FuzzyDBSCAN(data, eps, minPtsMin, minPtsMax);
    for (int cluster : clustering) {
        cout << cluster << " ";
    }
    cout << endl;
    return 0;
}

